src/Services/QueuedJobService.php

```diff
@@ -1533,6 +1579,271 @@ private function addJobHandlersToLogger(LoggerInterface $logger, QueuedJob $job,
+            ->sort('ID', 'ASC');
+
+        // Provide an option to customise the list of jobs that are eligible for a job retry
+        $this->extend('updateRetryJobsList', $autoRetryJobs);
+
+        $jobConfigs = [];
+        $jobClasses = $autoRetryJobs->columnUnique('Implementation');
+
+        foreach ($jobClasses as $jobClass) {
+            // Skip invalid job data (orphaned data)
+            if (!class_exists($jobClass)) {
+                continue;
+            }
+
+            $maxRetryAttempts = (int) Config::inst()->get($jobClass, 'max_retry_attempts');
+
+            // Job is not configured to be retried
+            if ($maxRetryAttempts <= 0) {
+                continue;
+            }
+
+            $initialRetryDelay = (int) Config::inst()->get($jobClass, 'initial_retry_delay');
+            $retryFalloffMultiplier = (float) Config::inst()->get($jobClass, 'retry_falloff_multiplier');
+            $retryFalloffMultiplierVariance = (float) Config::inst()->get(
+                $jobClass,
+                'retry_falloff_multiplier_variance'
+            );
+
+            // Sanitise values so we don't have to deal with edge cases
+            $initialRetryDelay = max(0, $initialRetryDelay);
+            $retryFalloffMultiplier = max(1, $retryFalloffMultiplier);
+            $retryFalloffMultiplierVariance = max(0, $retryFalloffMultiplierVariance);
+
+            $jobConfigs[$jobClass] = [
+                $maxRetryAttempts,
+                $initialRetryDelay,
+                $retryFalloffMultiplier,
+                $retryFalloffMultiplierVariance,
+            ];
+        }
+
+        // We couldn't find any job candidates to retry
+        if (!$jobConfigs) {
+            return;
+        }
+
+        $retryJobIDs = [];
+        $remainingLimit = $jobRetryLimit;
+
+        foreach ($jobConfigs as $jobClass => $retryConfig) {
+            // We've reached the number of jobs we are allowed to retry in this health check run
+            if ($remainingLimit <= 0) {
+                break;
+            }
+
+            [
+                $maxRetryAttempts,
+            ] = $retryConfig;
+
+            $jobIDs = $autoRetryJobs
+                ->filter([
+                    'Implementation' => $jobClass,
+                    'RetryCount:LessThan' => $maxRetryAttempts,
+                ])
+                ->limit($remainingLimit)
+                ->column('ID');
+
+            // Add the found jobs to our list so we can retry these jobs
+            $retryJobIDs = array_merge($retryJobIDs, $jobIDs);
+
+            // Subtract found jobs from the remaining limit
+            $remainingLimit -= count($jobIDs);
+        }
+
+        // We couldn't find any jobs to retry
+        if (!$retryJobIDs) {
+            return;
+        }
+
+        $jobsToRetry = $autoRetryJobs->byIDs($retryJobIDs);
+
+        /** @var QueuedJobDescriptor $jobDescriptor */
+        foreach ($jobsToRetry as $jobDescriptor) {
```

Comment:
Why is this done in a separate loop to the `foreach ($jobConfigs as $jobClass => $retryConfig) {` loop above? We could skip collecting an array of IDs and making this `byIDs()` call by just scheduling the jobs right away in that initial loop.

This would mean if there's a _lot_ of jobs to retry, our SQL query will (on average, assuming lots of jobs types exist with retries) have less records to iterate over at a time (since we're just doing one class at a time, not all at once).

The difference in practice is probably minimal though, so even just "this is cleaner code" is probably a good enough reason to keep these loops separate (especially if they're extracted into separate `private` methods as suggested above).